### ISSUE TITLE
Integrating external dependencies (configure/make)

### ISSUE THREAD

**Issue by TheQwertiest** [mesonbuild/meson#5001]
I'm trying to migrate to Meson build from our existing configure/make build system (the pain...). I've successfully rewritten the build logic for most of component, but stumbled on integrating external dependencies, which are compiled from source files.

To avoid possible confusion, I will provide the sample with comments marking the questions, which will be listed after the sample.

Current build scenario:
1. External dependencies are downloaded via dependency manager in `_result` directory.
2. `component` configuration & building.

<details><summary>Component folder structure</summary><p>

```
# meson build
_build
# contains external dependencies downloaded via step 1.
# also used as --prefix for meson.
_result
  /bin
  /lib
    /dependency_1.a
    /dependency_2.so
  /include
    /dependency_1/
    /dependency_2/
    /dependency_3/
  /src
    /dependency_3/
component
  /meson.build
  /proj_1
    /meson.build
  /subprojects
    /dependency_1
    /dependency_2
    /dependency_3
```
</p></details><br>

All of the questions are in `component/subprojects/dependency_3/meson.build` all other files are mostly for reference.

<details><summary>component/meson.build (no questions here, defines vars used in other meson.builds)</summary><p>

```Meson
project('component', ['cpp'], default_options: ['prefix=' + meson.source_root()/'../_result', 'layout=flat'])

path_prefix = get_option('prefix')
path_libdir = path_prefix/get_option('libdir')
path_includedir = path_prefix/get_option('includedir')

path_rpath = '$ORIGIN/..'/get_option('libdir')
path_build_rpath = ';'.join([path_rpath, path_libdir])

# Meson requires relative paths for includes, so create a symbolic link to ../${LAST_DIR_FROM_PREFIX} if needed
prefix_split = path_prefix.split('/')
path_prefix_relative = '..'/prefix_split[prefix_split.length() - 1]

dirs_include = [include_directories(path_prefix_relative/get_option('includedir'), is_system: true)]

cpp = meson.get_compiler('cpp')

subdir('project_1')
```
</p></details><br>

<details><summary>component/project_1/meson.build (no questions here, used to illustrate dependency handling)</summary><p>

```Meson
src = [...]
cxxflags = [...]

lib_project_1 = library('project_1', src,
      cpp_args: cxxflags,
      include_directories: dirs_include,
      dependencies: [subproject('dependency_1').get_variable('dep_dependency_1'),
                             subproject('dependency_3').get_variable('dep_dependency_3')],
      install: true,
      build_rpath: path_build_rpath,
      install_rpath: path_rpath
  )
```
</p></details><br>

<details><summary>component/subprojects/dependency_1/meson.build (no questions here, used to illustrate dependency handling)</summary><p>

```Meson
project('dependency_1', 'cpp')

cpp = meson.get_compiler('cpp')

path_libdir = get_option('prefix')/get_option('libdir')

dep_dependency_1 = declare_dependency(dependencies: [cpp.find_library('dependency_1', dirs: path_libdir)])
```
</p></details><br>

<details><summary>component/subprojects/dependency_2/meson.build (no questions here, used to illustrate dependency handling)</summary><p>

```Meson
project('dependency_2', 'cpp')

cpp = meson.get_compiler('cpp')

path_libdir = get_option('prefix')/get_option('libdir')

dep_dependency_2 = declare_dependency(dependencies: [cpp.find_library('dependency_2', dirs: path_libdir), subproject('dependency_1').get_variable('dep_dependency_1')] )
```
</p></details><br>

<details><summary>component/subprojects/dependency_3/meson.build (all question marks are here)</summary><p>

```Meson
project('dependency_3', 'cpp')

# Note: this code is not entirely valid, but rather illustrates how I though it would be working

cpp = meson.get_compiler('cpp')

path_prefix = get_option('prefix')
path_libdir = path_prefix/get_option('libdir')
path_srcdir = path_prefix/'src'

path_dependency_3_src = path_srcdir/'dependency_3'/'dependency_3'

# Question 1
env = environment()
env.set('CXX', ' '.join(meson.get_compiler('cpp').cmd_array()))
env.set('CXXFLAGS', ' '.join(get_option('cpp_args')))
env.set('LDFLAGS', ' '.join(get_option('cpp_link_args')))

# Question 2
if meson.is_cross_build()
  # e.g. 'arm-linux-gnueabihf-'
  cross_base = meson.get_cross_property('cross_base')
else 
  cross_base = ''
endif

env.set('LD', cross_base + 'ld')
env.set('AR', cross_base + 'ar')
env.set('RANLIB', cross_base + 'ranlib')
env.set('STRIP', cross_base + 'strip')
env.set('OBJCOPY', cross_base + 'objcopy')

cfg_opts = ['--prefix=' + path_prefix] 

# Question 3
run_command(path_dependency_3_src /'configure', cfg_opts, env: env)
lib_dependency_3 = custom_target(
  'lib_dependency_3', 
  command: [find_program('make'), '-C', path_dependency_3_src /'Makefile'],
  input: [path_dependency_3_src /'Makefile.conf']
  dependencies: [subproject('dependency_1').get_variable('dep_dependency_1')],
  output: 'dependency_3.so',
  build_by_default: true
)

dep_dependency_3  = declare_dependency(dependencies: [lib_dependency_3])
```
</p></details><br>

**Questions**
1. Is there a way to avoid this with the current Meson? It's a lot of boiler-plate code, which has to be duplicated for every dependency that has to be built from sources. This boiler-plate code duplication can be completely avoided by extracting it to external makefile in both CMake and Make. https://github.com/mesonbuild/meson/pull/3557 might help with that if we are allowed to include `meson.build` from outside of `subprojects` directory, but it does not seem that that PR got much traction and/or support from devs.
2. Is there a proper way to pass currently used toolchain info to external build systems (e.g. Make/Cmake) ? Some values (e.g. CXX, CXXFLAGS) can be fetched and/or deduced with various functions. But not for the others (LD, AR, STRIP, RANLIB). This prevents from making proper cross-builds and native builds with `native-file` specified. The only workaround I could find is defining additional `cross_base` value in `cross-file`, but even that won't work, since there is no way to get values from `native-file` (i.e. `meson.get_native_property()`)
**Solved**: https://mesonbuild.com/Reference-manual.html#find_program (https://github.com/mesonbuild/meson/issues/2005)
3. How can I integrate `custom_target` into meson build? `declare_dependency` does not consider the output from `custom_target` to be a valid dependency target. Currently I can only build dependencies externally (outside Meson build system), which is a really dirty hack. There is somewhat related https://github.com/mesonbuild/meson/issues/4908, but I'm not sure if it will be applicable in this scenario.

There is https://github.com/mesonbuild/meson/pull/4321, which might somewhat help, but it seems that PR is rather stale (i.e. not being worked upon) and it does not seem to address Question 2.

Any help regarding this issues would be greatly appreciated (it's a show-stopper for migrating to Meson =( ).

**Comment by TheQwertiest** [comment]
Related question: https://github.com/mesonbuild/meson/issues/4872 - it's a bit different, since in my case sources are not inside `subprojects` (of course, it could be solved by creating symlinks, but that's feels like a dirty hack).

**Comment by TheQwertiest** [comment]
Another related question: https://github.com/mesonbuild/meson/issues/4316 - it seeks for a generic solution (which might be implemented somewhen in the future).

**Comment by TheQwertiest** [comment]
Sifted through github repositories a bit: it seems that external dependencies in projects that use Meson are either fetched with Wrap (and built later in Meson) or built by external tools before invoking Meson. Examples of the latter includes: configure > CMake { build dependencies, invoke Meson } (yuck), configure > automake { build dependencies, invoke Meson } (double-minus-yuck). Both of these examples feel like using Meson just for the sake of using Meson, since you still have to configure and write dependency relations via another build system...

**Comment by mensinda** [comment]
I hope that #4969 will solve this problem for CMake dependencies. However, implementing the same for Autotools will be a lot harder, since there is no build system introspection (at least that I know of).

**Comment by TheQwertiest** [comment]
@mensinda Regretfully, I have neither of those - we are using configure (that is plain configure.sh) + plain makefiles (yeah, I know). And while it would be possible to convert some subset of dependencies to Meson or Cmake, it's not possible for all of them.
And I doubt that having to wrap plain makefile projects into CMake and then invoking that from Meson would help me in advocating the transfer to Meson :D

**Comment by mensinda** [comment]
In that case, I only see two options to resolve this:

1. Build a Makefile parser in meson and somehow extract libraries and targets
2. Build all dependencies at meson configure time and find the generated libraries

I don't think that any of these options are particularly viable for meson (at least as far as I can tell, I could be wrong here):

The first option would require a complete Makefile parser and additional logic to extract abstract build targets like libraries and executables. The second option could be implemented, but I don't think that building dependencies at configure time is any better than building the dependencies before invoking meson.

Again I could be missing something here, but this won't be something that can be implemented easily.

**Comment by TheQwertiest** [comment]
Yeah, I don't think that 1. is feasible, since Makefile might contain a lot of magic (invoking bash for starters) and does not provide some sort of dependency list output.

Regarding 2. - I thought that it could be handled via `custom_target` (granted I have no idea about inner workings of meson (or ninja)): `custom_target` has `input`, `output` and `depends`, so for a `target` that depends on `custom_target` it would be possible to make a relation `depends` > `input` > `output` > `target`, which could be resolved at build time.

> I don't think that building dependencies at configure time is any better than building the dependencies before invoking meson.

Not sure if wrapping meson build in CMake (or other build system) is better though.

**Comment by mensinda** [comment]
The Problem with `custom_target` would be that you have to specify the exact output libraries of the dependency. This can be easily done if you only have one system, but quickly falls apart for multiple systems (windows with different compilers have different locations/formats of dynamic libraries for instance, also import libraries, etc.). It gets even worse with custom makefiles that can do _anything_.

In theory, this could be implemented, but has to account for all the problems described above, from my experience with my current CMake subprojects PR alone, I can tell that this would be a nightmare to maintain (for either meson or the project in question).

**Comment by TheQwertiest** [comment]
> The Problem with `custom_target` would be that you have to specify the exact output libraries of the dependency...

That would be at least better than nothing that we have currently now =)
I don't think that our project is the only one who has `legacy` dependencies that they can't get rid of (or at least I hope so)...

**Comment by mensinda** [comment]
If you don't mind `custom_target`, you could create a python script (python is guaranteed to work with meson, since python is required to run meson) that configures and builds your dependencies. You should then be able to use the output of `custom_target` with either `link_with` or `link_args`.

Something like:

```meson
prog    = find_program('yourScript.py')
out_lib = custom_target('gen_dep',
  output:  ['libA.so', 'libB.so'],
  command: [prog, '@OUTPUT0@', '@OUTPUT1@']
)

executable('yourExe', [], link_args: [out_lib])
```
Note: I would consider this to be an awful hack :)

I also haven't tested this code and I am not sure if this will work. If it doesn't then it should be possible to make a PR to implement this in meson. Again, there are multiple things that can break here!

**Comment by dcbaker** [comment]
I've been thinking about this a bunch recently. One idea I've had is to write a module for this, with a rough API such as:
```meson
mk = import("make").init('projectname')
mk.configure(['--list', 'of', '--options'])  # assumes a `./configure` or `./autogen.sh`
libfoo = mk.build('foo.so', name : 'foo.so', path : 'src/lib/', include : 'src/include/foo')
```

Then you can use the target as a normal meson library. the configure set would be invoked during meson setup, and the build step would be invoked by ninja. There are a bunch of things to consider:
passing environment through (such as pkg-config or cmake overrides), passing -j and -l through correctly, a lot of stuff I haven't thought about.

On the other hand make is a pretty simple syntax, so it might be possible to just run configure and then parse the makefiles to build an AST.

**Comment by mensinda** [comment]
> On the other hand make is a pretty simple syntax, so it might be possible to just run configure and then parse the makefiles to build an AST.

Yes, the syntax of Makefiles is pretty simple. However the main problem would be to interpret the result. Which targets should be built, given a full library name (`libfoo.so`, `libfoo.dll`, `libfoo.dll.a`) what is the base name of the target (`foo` in this case). You also would have to filter out `clean` targets and other stuff the user might have written manually.

Granted, this are all problems that can be solved, I am just saying that this will be a lot of work to get right.

That beeing sait, this would be the most reliable approach once it is implemented. This is also how I have chosen implemented #4969, so it should also work for Makefiles in theory. Although I had it much easier since CMake provides an introspection API.

**Comment by TheQwertiest** [comment]
> You should then be able to use the output of `custom_target` with either `link_with` or `link_args`

Regretfully, this is not currently possible - it's blocked by https://github.com/mesonbuild/meson/issues/4908

> One idea I've had is to write a module for this, with a rough API such as...

That would be the ideal solution for most cases! It should take into account issues raised by my Q2 though. And it also needs `depends` argument and ability to output multiple files (that is a single invocation of make might produce multiple libraries).

The `custom_target` workaround would still be needed for custom build systems, such as Boost's jam or Mozilla's mach.

Does anyone know of any workarounds that could be used in the current Meson trunk in the mean time?

**Comment by mensinda** [comment]
> Does anyone know of any workarounds that could be used in the current Meson trunk in the mean time?

Again, I haven't tried this, but you could do the `custom_target()` like described as above, then use `declare_dependency()` to create a dependency object for your build target.

In `declare_dependency` or `executable` you would then provide the kwarg `link_args` with a manual path (as in a hardcoded string) to the generated library.

**WARNING:** Even if this works, this is a very awful hack!

**Comment by TheQwertiest** [comment]
> In `declare_dependency` or `executable` you would then provide the kwarg `link_args` with a manual path (as in a hardcoded string) to the generated library.

This does not create a relation (that is dependency) to `custom_target` (at least it didn't for me). If i just pass `-lpathToMyLib` it will try to find the lib (and fail), but won't actually invoke the custom target :\

> **WARNING:** Even if this works, this is a very awful hack!

I don't have much a choice here (aside from not using Meson). And as I've said before, `custom_target` is the only way to build smth with custom build systems (which do not have configure and/or make and thus can't be used with potential Make/Cmake meson module).

**Comment by TheQwertiest** [comment]
Regarding Question 2: I found the answer https://mesonbuild.com/Reference-manual.html#find_program (https://github.com/mesonbuild/meson/issues/2005)
Question 1 and Question 3 are still active.

**Comment by Ericson2314** [comment]
I wish there was a way to to just guess at configure time, and then turn all configure tests into unit tests. So you could "leap" guessing which dependencies will be resolved against the external stuff, have ninja call make, then "look" after the fact that all the configure time assumptions are in fact true.

**Comment by bonzini** [comment]
> This does not create a relation (that is dependency) to custom_target

There is also `link_depends`, so perhaps you could add `link_depends` support to `declare_dependency`. It's simpler than #4908.

**Comment by TheQwertiest** [comment]
> perhaps you could add `link_depends` support to `declare_dependency`

That would be still a (dirty) workaround, since it'll require to specify dependency twice, e.g.:
```meson
lib = custom_target(...)
lib_my_dep = declare_dependency(
  link_with: lib.full_path(),
  link_depends: lib 
)
```
But, yeah, at first glance it looks pretty easy to implement.

Another alternative would be to add smth like `relative_path`, which will return the path relative to build dir.
```meson
lib = custom_target(..., output: lib_name)
lib_my_dep = declare_dependency(
  link_with: lib.relative_path() # this will create a dependency, since it would be equal to the `build_dir/lib_name` (which will be made a target for ninja)
)

**Comment by nioncode** [comment]
This will only work for the ninja backend though. You need an explicit dependency to the target that creates the output, otherwise there is no way to make sure that this target is actually built before its output is consumed.

The proper way is to implement linking against custom targets.

**Comment by TheQwertiest** [comment]
Question 1 could be solved by https://github.com/mesonbuild/meson/pull/5209.
Question 3 is partially solved by https://github.com/mesonbuild/meson/pull/5103 and will be fully resolved by https://github.com/mesonbuild/meson/pull/5161.

**Comment by bonzini** [comment]
Closing as there is now support for external_project, plus questions 2 and 3 are solved. The remaining one is a dup of #375.
